### PR TITLE
Inject and configure Develocity Maven extension or Gradle plugin

### PR DESCRIPTION
**Pull request by pshevche**
## Summary
1. Update user-facing product references to "Develocity" instead of "Gradle Enterprise". All code references (e.g., variable names) will be updated as we change the injection logic.
2. Embed a Develocity extension instead of the GE extension.
3. Configure either the GE or the DV extension when found.

### CODE REVIEW

**Comment by pshevche** [comment]
❗️❗️❗️Blocked from merging until Develocity extension 1.21 is released❗️❗️❗️

File: agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java [hunk 1]
```diff
@@ -248,6 +254,13 @@ private String getMavenInvocationArgs(BuildRunnerContext runner) {
         return "-Dmaven.ext.class.path=" + asClasspath(extensionJars) + " " + asArgs(sysProps);
     }
 
+    private static boolean hasNoDevelocityExtensionApplied(BuildRunnerContext runner, MavenExtensions extensions) {
```
Comment (clayburn) [inline agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java:257]: The naming of `hasNoDevelocityExtensionApplied` confused me slightly. I only expected it to check for the `develocity-maven-extension`, not for the `gradle-enterprise-maven-extension` or a custom extension. That said, I can't think of a better name either that can encompass all three of these.

File: agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java [hunk 2]
```diff
@@ -207,32 +210,35 @@ private File getInitScriptsDir() {
...
-                : new File(gradleUserHomeOverride);
+            ? new File(System.getProperty("user.home", System.getenv("USER_HOME")), ".gradle")
+            : new File(gradleUserHomeOverride);
     }
 
     private String getMavenInvocationArgs(BuildRunnerContext runner) {
-        List<File> extensionJars = new ArrayList<File>();
-        List<String> sysProps = new ArrayList<String>();
+        List<File> extensionJars = new ArrayList<>();
+        List<String> sysProps = new ArrayList<>();
```
Comment (clayburn) [inline agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java:219]: `agent` is compiled with a Java 6 toolchain in [CI](https://github.com/etiennestuder/teamcity-build-scan-plugin/blob/main/.github/workflows/build-verification.yml#L31). Unfortunately, the diamond operator can't be used then.

```suggestion
        List<File> extensionJars = new ArrayList<File>();
        List<String> sysProps = new ArrayList<String>();
```

File: agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java [hunk 3]
```diff
@@ -275,7 +288,7 @@ private MavenExtensions getMavenExtensions(BuildRunnerContext runner) {
             return MavenExtensions.empty();
         }
 
-        final List<File> searchLocations = new ArrayList<File>();
+        final List<File> searchLocations = new ArrayList<>();
```
Comment (clayburn) [inline agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java:291]: Java 6 / Diamond Operator

```suggestion
        final List<File> searchLocations = new ArrayList<File>();
```

File: agent/service-message-maven-extension/src/main/java/nu/studer/teamcity/buildscan/agent/servicemessage/BuildScanServiceMessageDevelocityListener.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+)
+public final class BuildScanServiceMessageDevelocityListener extends BuildScanServiceMessageListener implements DevelocityListener {
+
+    @Override
+    public void configure(DevelocityApi api, MavenSession session) {
+        super.configure(new BuildScanApiAdapter() {
+            @Override
+            public void publishOnDemand() {
+                // publication is controlled via a system property in this case
+                api.getBuildScan().publishing(p -> p.onlyIf(ctx -> false));
```
Comment (clayburn) [inline agent/service-message-maven-extension/src/main/java/nu/studer/teamcity/buildscan/agent/servicemessage/BuildScanServiceMessageDevelocityListener.java:25]: Is this the same as publish on demand? Or does `p.onlyIf(ctx -> false)` mean that a build scan would never be published, even if `-Dscan` is passed?

**Review by clayburn — CHANGES_REQUESTED** [review]

File: agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java (hunk 2 above)
Comment (pshevche) [reply agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java]: Argh, right. Changed in https://github.com/etiennestuder/teamcity-build-scan-plugin/pull/373/commits/90f96034f60ab2301893309222081f61a13f25fc

File: agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java (hunk 3 above)
Comment (pshevche) [reply agent/src/main/java/nu/studer/teamcity/buildscan/agent/BuildScanServiceMessageInjector.java]: Changed in https://github.com/etiennestuder/teamcity-build-scan-plugin/pull/373/commits/90f96034f60ab2301893309222081f61a13f25fc

File: agent/service-message-maven-extension/src/main/java/nu/studer/teamcity/buildscan/agent/servicemessage/BuildScanServiceMessageDevelocityListener.java (hunk 4 above)
Comment (pshevche) [reply agent/service-message-maven-extension/src/main/java/nu/studer/teamcity/buildscan/agent/servicemessage/BuildScanServiceMessageDevelocityListener.java]: This is now removed in favor of the public adapters library: https://github.com/etiennestuder/teamcity-build-scan-plugin/pull/373/commits/08f2e9924c13e98aa86775b63179ea0d298289f9

See https://github.com/gradle/develocity-agent-adapters

File: src/main/java/nu/studer/teamcity/buildscan/connection/DevelocityConnectionConstants.java [hunk 5]
```diff
@@ -1,74 +1,74 @@
...
 
     public static final String GRADLE_PLUGIN_REPOSITORY_URL = "gradlePluginRepositoryUrl";
-    public static final String GRADLE_ENTERPRISE_URL = "gradleEnterpriseUrl";
+    public static final String DEVELOCITY_URL = "gradleEnterpriseUrl";
     public static final String ALLOW_UNTRUSTED_SERVER = "allowUntrustedServer";
-    public static final String GE_PLUGIN_VERSION = "gradleEnterprisePluginVersion";
+    public static final String DEVELOCITY_PLUGIN_VERSION = "gradleEnterprisePluginVersion";
     public static final String CCUD_PLUGIN_VERSION = "commonCustomUserDataPluginVersion";
-    public static final String GE_EXTENSION_VERSION = "gradleEnterpriseExtensionVersion";
+    public static final String DEVELOCITY_EXTENSION_VERSION = "develocityExtensionVersion";
```
Comment (clayburn) [inline src/main/java/nu/studer/teamcity/buildscan/connection/DevelocityConnectionConstants.java:14]: Was the value of this string changed intentionally? I notice others, such as `DEVELOCITY_PLUGIN_VERSION` were not changed.

These constants are eventually used by TC to persist the connection configuration to disk, so changing these could break the configuration of upgraders.

Comment (pshevche) [reply src/main/java/nu/studer/teamcity/buildscan/connection/DevelocityConnectionConstants.java]: You are right, I shouldn't have changed it. Reverted in https://github.com/etiennestuder/teamcity-build-scan-plugin/pull/373/commits/a58d23bfe53587904e44d9327fd563f7072041ed

**Review by clayburn — APPROVED** [review]

File: agent/service-message-maven-extension/pom.xml [hunk 6]
```diff
@@ -12,6 +12,13 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     </properties>
 
+    <repositories>
+        <repository>
+            <id>Develocity extension RC repository</id>
+            <url>https://repo.grdev.net/artifactory/public</url>
+        </repository>
```
Comment (etiennestuder) [inline agent/service-message-maven-extension/pom.xml:19]: @clayburn do we still need this?

File: agent/service-message-maven-extension/pom.xml [hunk 7]
```diff
@@ -68,6 +86,19 @@
...
+                <artifactId>maven-shade-plugin</artifactId>
+                <version>3.5.2</version>
+                <executions>
+                    <execution>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                    </execution>
+                </executions>
```
Comment (etiennestuder) [inline agent/service-message-maven-extension/pom.xml:100]: How come this is needed?

**Review by etiennestuder — CHANGES_REQUESTED** [review]

File: agent/service-message-maven-extension/pom.xml (hunk 6 above)
Comment (pshevche) [reply agent/service-message-maven-extension/pom.xml:19]: This is not needed anymore, it was from the time when the extension was compiled against the RC version.

File: agent/service-message-maven-extension/pom.xml (hunk 7 above)
Comment (pshevche) [reply agent/service-message-maven-extension/pom.xml:100]: This is for including the adapters library in the extension JAR.

File: agent/service-message-maven-extension/pom.xml (hunk 6 above)
Comment (clayburn) [reply agent/service-message-maven-extension/pom.xml]: I've removed it

**Pull request merged**